usr/sbin/rear

```diff
@@ -510,41 +518,122 @@ fi
 # so that ReaR functions for actually intended user messages can use fd7 and fd8
-# to show messages to the user regardless where to STDOUT and STDERR are redirected
-# and fd6 to get input from the user regardless where to STDIN is redirected:
+# to show messages to the user regardless where STDOUT and STDERR are redirected
+# and fd6 to get input from the user regardless where STDIN is redirected:
 if ! source $SHARE_DIR/lib/_input-output-functions.sh ; then
     echo -e "ERROR: BUG in $PRODUCT\nFailed to source $SHARE_DIR/lib/_input-output-functions.sh\nPlease report it at $BUG_REPORT_SITE" >&2
     exit 1
 fi
 
 # Via source usr/share/rear/lib/_input-output-functions.sh
-# those exit tasks are now set to be executed (via bash EXIT trap) in the following order:
+# those final exit tasks are now set to be executed (via bash EXIT trap) in the following order:
 # "(( EXIT_FAIL_MESSAGE )) && echo '${MESSAGE_PREFIX}$PROGRAM $WORKFLOW failed, check $RUNTIME_LOGFILE for details' 1>&8"
 # "exec 8>&-"
 # "exec 7>&-"
 # "exec 6<&-"
 
-# Used to determine whether TMPDIR has been changed in user config.
-# After the `mktemp` below, changing TMPDIR does not have the intended effect.
-# Save the current value to detect changes.
-saved_tmpdir="${TMPDIR-}"
+# Set RECOVERY_MODE when we are running inside the rescue/recovery system.
+# RECOVERY_MODE is a boolean variable (cf. conf/default.conf) because we need it below
+# to set TMPDIR to ReaR's TMP_DIR only when we are not running inside the recovery system
+# but we do not yet have lib/global-functions.sh (which contains is_false) sourced here.
+# In the recovery system /etc/rear-release is unique (it does not exist otherwise)
+# see build/default/970_add_rear_release.sh that adds it to identify the rescue environment:
+test -e "/etc/rear-release" && RECOVERY_MODE='y' || RECOVERY_MODE=''
+readonly RECOVERY_MODE
+
+# Create a sufficiently safe temporary working area BUILD_DIR for ReaR and
+# additionally set TMPDIR to TMP_DIR within BUILD_DIR for programs that are called by ReaR
+# (the latter only when ReaR runs on the original system i.e. when we are not in RECOVERY_MODE):
 if test "$WORKFLOW" != "help" ; then
-    # Prepend temporary working area (BUILD_DIR) removal exit task
-    # so it gets executed directly before the above listed exit tasks.
-    # This must be done before the first possible call of the 'Error' function
-    # otherwise we may error out but leave the build area behind:
+    # ReaR fails when TMPDIR is a relative directory, see
+    # https://github.com/rear/rear/pull/3168#issuecomment-1999996162
+    # so we set TMPDIR to its resolved absolute path.
+    # Ignore when 'readlink -e' fails (e.g. it won't fail when TMPDIR is a regular file)
+    # and verify afterwards that the 'readlink -e' result is a directory:
+    export TMPDIR="$( readlink -e "${TMPDIR-}" || true )"
+    if ! test -d "$TMPDIR" ; then
```

Comment:
I was curious how this will work when `TMPDIR` is unset or empty initially. Maybe add a comment that TMPDIR must be already set here, which is performed when the defaults are sourced?

Reply:
FYI:
How it fails with current code when TMPDIR is empty:
```
# export TMPDIR=" "

# usr/sbin/rear -d mkrescue
...
ERROR: TMPDIR '' is no directory
...

# echo "'$( readlink -e "$TMPDIR" )'"
''
```

Reply:
Fixed by
https://github.com/rear/rear/commit/96867121a3cdf0596d1f0d9ea0458009fa2703c3